flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java

```diff
@@ -195,24 +212,39 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
 
     @Override
     public void onNewResourcesAvailable() {
-        evaluateRescaling();
+        rescaleManager.onChange();
+        initializeFailedCheckpointCountdownIfUnset();
     }
 
     @Override
     public void onNewResourceRequirements() {
-        evaluateRescaling();
-    }
-
-    private void evaluateRescaling() {
         rescaleManager.onChange();
-        rescaleManager.onTrigger();
+        initializeFailedCheckpointCountdownIfUnset();
     }
 
     @Override
-    public void onCompletedCheckpoint() {}
+    public void onCompletedCheckpoint() {
+        triggerPotentialRescale();
+    }
 
     @Override
-    public void onFailedCheckpoint() {}
+    public void onFailedCheckpoint() {
+        if (this.failedCheckpointCountdown != null
+                && this.failedCheckpointCountdown.decrementAndGet() <= 0) {
+            triggerPotentialRescale();
+        }
+    }
+
+    private void triggerPotentialRescale() {
+        rescaleManager.onTrigger();
+        this.failedCheckpointCountdown = null;
+    }
```

Comment:
Are we sure these `onX...` methods can't be called concurrently? To not get a NPE when calling `  && this.failedCheckpointCountdown.decrementAndGet() <= 0)`

Reply:
Synchronization should have been covered by the `CheckpointStatsListener` wrapper implementation in `AdaptiveScheduler`.